rs-matter/src/data_model/core.rs

```diff
@@ -15,142 +15,759 @@
+
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+
+    async fn report_data(
+        &self,
+        id: u32,
+        node_id: u64,
+        rx: &[u8],
+        tx: &mut [u8],
+        exchange: &mut Exchange<'_>,
+    ) -> Result<bool, Error>
+    where
+        T: DataModelHandler,
+    {
+        let mut wb = WriteBuf::new(tx);
+
+        let req = SubscribeReq::from_tlv(&get_root_node_struct(rx)?)?;
+        let req = ReportDataReq::Subscribe(&req);
+
+        let metadata = self.handler.lock().await;
+
+        let accessor = exchange.accessor()?;
+
+        {
+            let node = metadata.node();
+            let mut attrs = node.read(&req, None, &accessor).peekable();
 
-                    driver.complete(req).await?;
+            loop {
+                let more_chunks = req
+                    .respond(&self.handler, Some(id), &mut attrs, &mut wb, false)
+                    .await?;
+
+                exchange.send(OpCode::ReportData, wb.as_slice()).await?;
+
+                if !Self::recv_status(exchange).await? {
+                    info!("Subscription {node_id:x}::{id} removed during reporting");
+                    return Ok(false);
+                }
+
+                if !more_chunks {
+                    break;
                 }
             }
         }
 
+        Ok(true)
+    }
+
+    async fn rx_buffer(&self, exchange: &mut Exchange<'_>) -> Result<Option<B::Buffer<'a>>, Error> {
+        let buffer = self.tx_buffer(exchange).await?;
+
+        if let Some(mut buffer) = buffer {
+            let rx = exchange.rx()?;
+
+            buffer.clear();
+            buffer.extend_from_slice(rx.payload()).unwrap();
+
+            exchange.rx_done()?;
+
+            Ok(Some(buffer))
+        } else {
+            Ok(None)
+        }
+    }
+
+    async fn tx_buffer(&self, exchange: &mut Exchange<'_>) -> Result<Option<B::Buffer<'a>>, Error> {
+        if let Some(mut buffer) = self.buffers.get().await {
+            buffer.resize_default(MAX_EXCHANGE_TX_BUF_SIZE).unwrap();
+
+            Ok(Some(buffer))
+        } else {
+            Self::send_status(exchange, IMStatusCode::Busy).await?;
+
+            Ok(None)
+        }
+    }
+
+    async fn recv_status(exchange: &mut Exchange<'_>) -> Result<bool, Error> {
+        let rx = exchange.recv().await?;
+        let opcode = rx.meta().proto_opcode;
+
+        if opcode == OpCode::StatusResponse as u8 {
+            let resp = StatusResp::from_tlv(&get_root_node_struct(rx.payload())?)?;
+
+            if resp.status == IMStatusCode::Success {
+                Ok(true)
+            } else {
+                warn!(
+                    "Got status response {:?}, aborting interaction",
+                    resp.status
+                );
+
+                drop(rx);
+                exchange.acknowledge().await?;
+
+                Ok(false)
+            }
+        } else {
+            Err(ErrorCode::Invalid.into())
```

Comment:
should we add some levels of tracing (even if debug or info) as to not lose the actual value? An error of `invalid opcode` would be harder to debug than having a `Invalid opcode: received 0xF0 instead of 0x01`

Reply:
I would suggest doing this as a subsequent PR, which would be targeted specifically at improving the `Error` struct, and the `ErrorCode` enum.

We are leaking such information at a lot of places. Other cases: attribute not found, invalid command and so on.
Addressing this is IMO out of scope for this PR, even if the `InvalidOpcode` enum _was_ introduced as part of this PR (which is anyway gigantic already). I'm just following the existing status quo w.r.t. error codes, even if that's suboptimal.

Reply:
As for tracing, it is not ideal as the error should itself carry enough information, but yeah - I've added a `warn!` there.